autovalet_lane_detection/scripts/autovalet_lane_detector.py

```diff
@@ -0,0 +1,186 @@
+            self.publishLaneCloud(lane_cloud)
+
+    def center_line_detection(self, img):
+        # colorspace transformation
+        rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+        hls = cv2.cvtColor(img, cv2.COLOR_BGR2HLS).astype(np.float)
+
+        # hls thresholding for yellow
+        lower = np.array([12.0, 0.18*255, 0.90*255], dtype=np.uint8)
+        upper = np.array([30.0, 1.00*255, 1.00*255], dtype=np.uint8)
+        mask  = cv2.inRange(hls, lower, upper)
+        th    = cv2.bitwise_and(rgb, rgb, mask=mask).astype(np.uint8)
+        th    = cv2.cvtColor(th, cv2.COLOR_HLS2RGB)
+        th    = cv2.cvtColor(th, cv2.COLOR_RGB2GRAY)
+        th    = cv2.GaussianBlur(th, (5, 5), 0)
+        _, th = cv2.threshold(th, 60, 255, cv2.THRESH_BINARY)
+	th    = cv2.morphologyEx(th, cv2.MORPH_OPEN, (5,5))
+
+        # extract ROI
+        mask = np.zeros_like(th)
+        mask[self.ROI_UPPER_Y:, :self.ROI_RIGHT_X] = 1
+        roi = cv2.bitwise_and(th, th, mask=mask)
+
+        # detect center line
+        minLineLength, maxLineGap = 130, 20
+        edges = cv2.Canny(roi, 100, 200, apertureSize=3)
+        lines = cv2.HoughLinesP(edges, 1, np.pi/180, 30, np.array([]), minLineLength, maxLineGap)
+
+        # possible scenarios: 1. new line 2. tracking
+        max_len, scenario, X1, Y1, X2, Y2 = 0, 0, 0, 0, 0, 0
+        if lines is not None: # 1. new line
+            for line in lines:
+                x1, y1, x2, y2 = line[0]
+                line_len = np.sqrt((x2-x1)**2 + (y2-y1)**2)
+                if max_len < line_len:
+                    max_len = line_len
+                    X1, Y1, X2, Y2 = x1, y1, x2, y2
+            scenario = 1
+            bb = (X1, Y1, np.abs(X2-X1), np.abs(Y2-Y1))
+            self.tracker = cv2.TrackerKCF_create()
+            self.tracker.init(rgb, bb)
+            self.lose_track_count = 0
+        elif self.tracker is not None:
+            success, bb = self.tracker.update(rgb)
+            x, y, w, h = [int(v) for v in bb]
+            if success: # 2. tracking
+                scenario = 2
+                X1, Y1, X2, Y2 = x, y, x+w, y-h
+                Y1 = min(Y1, img.shape[0] - 1)
+
+
+        # coordinates and slope
+        if scenario == 0:
+            return np.array([])
+        if X1 == X2: # vertical line
+            coordinates = [[X1, i] for i in range(Y2, Y1+1)]
+        else:
+            slope = 1.0 * (Y2 - Y1) / (X1 - X2) # assuming slope being positive, might have bug
+            if slope > 1:
+                coordinates = [[X1+i, int(Y1-i*slope)] for i in range(X2-X1)]
+            else:
+                coordinates = [[int(X1+i/slope), Y1-i] for i in range(Y1-Y2)]
+
+        return np.array(coordinates)
+
+    def line2cloud(self, depth, coordinates):
+        # ref: Subbu's code
+        u, v = coordinates[:, 0], coordinates[:, 1]
+        x    = (u - self.cx) / self.fx
+        y    = (v - self.cy) / self.fy
```

Comment:
write the matrix equation in the comments plz

Reply:
inserted some comments above.

Reply:
where's s in the math? also just put in the def of K? I'm dumb sry.

here i did it for you
`K = [fx, 0, cx; 0, fy, cy; 0, 0, 0]` 

is it all zeros in the bottom row?

Reply:
s would be in row 1 column 2 like this: 'K = [fx, s, cx;  0, fy, cy;  0, 0, 0]'. We assume s is zero in the math

Reply:
what??? nooo.... s is not zero, we substitute it with z